modules/internal/ChapelBase.chpl

```diff
@@ -546,17 +546,32 @@ module ChapelBase {
   //
   // min and max
   //
+  inline proc min(x: int(?w), y: int(w)) return if x < y then x else y;
+  inline proc max(x: int(?w), y: int(w)) return if x > y then x else y;
+
+  inline proc min(x: uint(?w), y: uint(w)) return if x < y then x else y;
+  inline proc max(x: uint(?w), y: uint(w)) return if x > y then x else y;
+
+  inline proc min(x: real(?w), y: real(w)) return if x < y then x else y;
+  inline proc max(x: real(?w), y: real(w)) return if x > y then x else y;
+
+  inline proc min(x: imag(?w), y: imag(w)) return if x < y then x else y;
+  inline proc max(x: imag(?w), y: imag(w)) return if x > y then x else y;
+
   inline proc min(x, y) return if x < y then x else y;
   inline proc max(x, y) return if x > y then x else y;
-  inline proc min(x, y)
-    where chpl_isSyncSingleAtomic(x) || chpl_isSyncSingleAtomic(y)
-  { compilerError("min() and max() are not allowed on sync/single/atomic arguments - apply readFE/readFF/read() to those arguments first"); }
-  inline proc max(x, y)
-    where chpl_isSyncSingleAtomic(x) || chpl_isSyncSingleAtomic(y)
-  { compilerError("max() and min() are not allowed on sync/single/atomic arguments - apply readFE/readFF/read() to those arguments first"); }
```

Comment:
I disagree.  Keep in mind that most references to syncs in a program are reads of the sync.  If you passed a sync int into a routine expecting an int, it would be a readFE of the sync int (which is why the operators behave this way -- they're not special in any way.  The I/O writeln() case is weird because it's ambiguous whether you want the program to block if the sync is empty or whether you want it to write out the sync's value and full/empty state.  The I/O read() case is similar -- should it block if the sync variable is already full?  I don't think min/max are at all the same in this regard.  Looking through the history, my understanding is that the reason this was done was because if you fed sync vars into the min()/max() routines, those routines were generic (so would take in the sync variables) and then refer to their arguments x and y multiple times each.  So they would end up blocking because the syncs/singles would be emptied on the first read (if they were not already) and then would block on the second refernce (which was triyng to return them).  This made programs deadlock if they called min()/max() on them.  I'm not sure why, at the time, we decided to make these operations illegal rather than explicitly defining them on syncs/singles in a way that wouldn't deadlock, but I think we mostly didn't give them sufficient time (and were arguably unduly wed to keeping them generic).  With fresh eyes this week, it seems obvious to me that min()/max() should support the standard typed overloads (the main point of this commit) and that syncs/singles shouldn't be special in any way w.r.t. these routines (which they now aren't).  

If it's not clear, the reason for keeping the atomic error case around is that the atomics will fall into the generic case otherwise (which seems useful to keep around) and will then fail to resolve the < or > applied to them, leading to a confusing error message.  It seemed better to promote the error message to the level a user would understand.  We could arguably define a min()/max() on atomics that applied a read() to its arguments, but that seems... weird.  We wouldn't do that in any other context (whereas the semantics for a reference to a sync/single have long been to use its default read).

"the write behavior" -- I see what you did there.  :)